src/controller.rs

```diff
@@ -171,9 +174,45 @@ impl AppState {
                 .into_iter()
                 .map(|u| Arc::new(RpcClient::new(get_http_url(u).expect("valid RPC url"))))
                 .collect(),
+            rpc_endpoint: endpoint.to_string(),
```

Comment:
I think this is already available on the client, something like:
```suggestion
           state.client.rpc().url()
```